src/Tasks/Microsoft.NET.Build.Tasks/AppHost.cs

```diff
@@ -99,56 +100,89 @@ private static int[] ComputeKMPFailureFunction(byte[] pattern)
                 {
-                    if (table[i] > -1)
+                    if (pattern[i] == bytes[m + i])
                     {
-                        m = m + i - table[i];
-                        i = table[i];
+                        if (i == pattern.Length - 1)
+                        {
+                            return m;
+                        }
+                        i++;
                     }
                     else
                     {
-                        m++;
-                        i = 0;
+                        if (table[i] > -1)
+                        {
+                            m = m + i - table[i];
+                            i = table[i];
+                        }
+                        else
+                        {
+                            m++;
+                            i = 0;
+                        }
                     }
                 }
+
+                return -1;
+            }
+            finally
+            {
+                if (accessor.SafeMemoryMappedViewHandle != null)
+                {
+                    accessor.SafeMemoryMappedViewHandle.ReleasePointer();
+                }
             }
-            return -1;
         }
 
-        private static void SearchAndReplace(byte[] array, byte[] searchPattern, byte[] patternToReplace, string appHostSourcePath)
+        private static void SearchAndReplace(
+            MemoryMappedViewAccessor accessor,
+            byte[] searchPattern,
+            byte[] patternToReplace,
+            string appHostSourcePath)
         {
-            int offset = KMPSearch(searchPattern, array);
-            if (offset < 0)
+            int position = KMPSearch(searchPattern, accessor);
+            if (position < 0)
             {
                 throw new BuildErrorException(Strings.AppHostHasBeenModified, appHostSourcePath, _placeHolder);
             }
 
-            patternToReplace.CopyTo(array, offset);
+            accessor.WriteArray(
+                position: position,
+                array: patternToReplace,
+                offset: 0,
+                count: patternToReplace.Length);
+
+            Pad0(accessor, searchPattern, patternToReplace, position);
+        }
+
+        private static unsafe void Pad0(MemoryMappedViewAccessor accessor, byte[] searchPattern, byte[] patternToReplace, int position)
+        {
+            byte* ptrMemoryMappedView = null;
+            accessor.SafeMemoryMappedViewHandle.AcquirePointer(ref ptrMemoryMappedView);
 
             if (patternToReplace.Length < searchPattern.Length)
             {
                 for (int i = patternToReplace.Length; i < searchPattern.Length; i++)
                 {
-                    array[i + offset] = 0x0;
+                    byte empty = 0x0;
+                    *(ptrMemoryMappedView + i + position) = empty;
                 }
             }
+
+            accessor.SafeMemoryMappedViewHandle.ReleasePointer();
```

Comment:
This should also be in a finally.

It might be nicer to put it all in one try block calling two extracted methods that take byte* bytes, int length.

Reply:
The diff might be harder to see. Is this good? I moved pointer acquiring and releasing to the caller method. Only one try-finally is used 

https://github.com/dotnet/sdk/pull/2293/commits/a1a4f04791aef72abd76b1a1b4eaa26e1ca17ff2